### ISSUE TITLE
Potential inconsistency in `dask.delayed` best practices

### ISSUE THREAD

**Issue by zzhengnan** [dask/dask#8135]
Hi team. I'm somewhat puzzled by the following tips from https://docs.dask.org/en/latest/delayed-best-practices.html; specifically, tips 2 and 3 seem to be at odds with tip 1.
> 1. Avoid calling delayed within delayed functions

> 2. Don’t call dask.delayed on other Dask collections

> 3. Avoid repeatedly putting large inputs into delayed calls

Regarding tip 2, one of the examples given in the docs is as follows. `part` is a delayed object, and yet is being called by a delayed function (`train`). Doesn't this violate tip 1?
```python
partitions = df.to_delayed()
delayed_values = [dask.delayed(train)(part)
                  for part in partitions]
```

Similarly, regarding the example for tip 3, `x` is a delayed object, which is then called by a delayed function. Same question as above.
```python
x = np.array(...)    # some large array
x = dask.delayed(x)  # delay the data once
results = [dask.delayed(train)(x, i)
           for i in range(1000)]
```

I would appreciate any clarification. Thanks!

**Comment by douglasdavis** [comment]
In your first code block: `delayed` is not called inside of a `delayed` function (which is discouraged); `delayed` has already been called by creating the list of `Delayed` objects in `df.to_delayed`, so you're really just passing existing `Delayed` objects (created by `df.to_delayed`) to a `delayed` function, and this is fine.

In your second code block, the "large input" (`x`) is passed to a `delayed` call only once, and this is OK. It would be a bad idea to write
```python
x = np.array(...)
results = [dask.delayed(train)(delayed(x), i) for i in range(1000)]
```
that would be repeatedly putting large inputs into `delayed` calls

Hope this helps!

**Comment by zzhengnan** [comment]
Thanks for the reply @douglasdavis. I'm afraid I'm still not getting it.

Regarding your comment below, I thought we shouldn't pass a Dask collection into a delayed function, [per the docs](https://docs.dask.org/en/latest/delayed-best-practices.html#don-t-call-dask-delayed-on-other-dask-collections)?
> you're really just passing existing Delayed objects (created by df.to_delayed) to a delayed function, and this is fine.

Specifically, the following example is marked as a usage example that should be discouraged.
![image](https://user-images.githubusercontent.com/17503044/133710483-746925b0-0e39-4425-a703-a53c69a26c87.png)

Could you help me understand what I'm missing. Thanks!

**Comment by jsignell** [comment]
Maybe this is the piece that you are missing: when you call `.to_delayed()` on a dask collection (like a dataframe) what you get back is a list of delayed objects. So it's no longer a collection.

**Comment by zzhengnan** [comment]
@jsignell Thanks for the clarification. 

I think my confusion may have stemmed from the fact that, per the first sentence on [this page](https://docs.dask.org/en/latest/custom-collections.html), I'm interpreting the term "dask collection" to include delayed objects _in addition to_ dask arrays and dataframes. However, it sounds like that interpretation is not appropriate for the delayed best practices page, where "dask collection" should be interpreted to _exclude_ delayed objects? Does that sound right?

To summarize the discussion, does the following look right?
```python
# Bad
delayed_function(dask_array)

# Bad, if `numpy_array` is really big
delayed_function(numpy_array)

# OK
delayed_function(dask_delayed_object)
```

**Comment by jsignell** [comment]
Yeah I think that's correct. There is definitely some confusion about terminology.

**Comment by zzhengnan** [comment]
Thanks for the confirmation.
